src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -146,24 +146,20 @@ public bool HasValue
         /// </returns>
         public static int Compare(StringSegment a, StringSegment b, StringComparison comparisonType)
         {
-            int minLength = Math.Min(a.Length, b.Length);
-            int diff = string.Compare(a.Buffer, a.Offset, b.Buffer, b.Offset, minLength, comparisonType);
-            if (diff == 0)
+            if (a.HasValue && b.HasValue)
             {
-                diff = a.Length - b.Length;
+                return a.AsSpan().CompareTo(b.AsSpan(), comparisonType);
+            }
+            else
+            {
+                CheckStringComparison(comparisonType); // must arg check before returning
+                return !a.HasValue ? (b.HasValue ? -1 : 0) : 1; // null sorts less than non-null, and two nulls sort as equal
```

Comment:
Not suggesting it here, but a question in general, does the runtime guarantee the value of a boolean to be 1 if it is true (unless one themselves deliberately attempts to unsafe cast a given byte to a bool?

Because in such case couldn't one do `return Unsafe.As<bool, byte>(ref a.HasValue) - Unsafe.As<bool, byte>(ref b.HasValue);` to make the return value branchless?

Reply:
Check the discussion over at https://github.com/dotnet/coreclr/pull/16138. Lots of good information regarding these kinds of optimizations over there.

Reply:
Oh wow that is an amazing resource, thank you so much :)